benchmarks/static_runtime/deep_wide_pt.h

```diff
@@ -31,6 +32,99 @@ struct DeepAndWide : torch::nn::Module {
   torch::Tensor mu_, sigma_, fc_w_, fc_b_;
 };
 
+// Implementation using native functions and pre-allocated tensors.
+// It could be used as a "speed of light" for static runtime.
+struct DeepAndWideFast : torch::nn::Module {
+  DeepAndWideFast(int num_features = 50) {
+    mu_ = register_parameter("mu_", torch::randn({1, num_features}));
+    sigma_ = register_parameter("sigma_", torch::randn({1, num_features}));
+    fc_w_ = register_parameter("fc_w_", torch::randn({1, num_features + 1}));
+    fc_b_ = register_parameter("fc_b_", torch::randn({1}));
+    allocated = false;
+    prealloc_tensors = {};
+  }
+
+  torch::Tensor forward(
+      torch::Tensor ad_emb_packed,
+      torch::Tensor user_emb,
+      torch::Tensor wide) {
+    torch::NoGradGuard no_grad;
+    if (!allocated) {
+      auto wide_offset = at::native::add(wide, mu_);
+      auto wide_normalized = at::native::mul(wide_offset, sigma_);
+      // Placeholder for ReplaceNaN
+      auto wide_preproc = at::native::clamp(wide_normalized, -10.0, 10.0);
+
+      auto user_emb_t = at::native::transpose(user_emb, 1, 2);
+      auto dp_unflatten = at::native::bmm_cpu(ad_emb_packed, user_emb_t);
+      // auto dp = at::native::flatten(dp_unflatten, 1);
+      auto dp = dp_unflatten.view({dp_unflatten.size(0), 1});
+      auto input = at::native::_cat_cpu({dp, wide_preproc}, 1);
+
+      // fc1 = torch::nn::functional::linear(input, fc_w_, fc_b_);
+      fc_w_t_ = torch::t(fc_w_);
+      auto fc1 = torch::addmm(fc_b_, input, fc_w_t_);
+
+      auto pred = at::native::sigmoid(fc1);
+
+      prealloc_tensors = {wide_offset,
+                          wide_normalized,
+                          wide_preproc,
+                          user_emb_t,
+                          dp_unflatten,
+                          dp,
+                          input,
+                          fc1,
+                          pred};
+      allocated = true;
+
+      return pred;
+    } else {
+      // Potential optimization: add and mul could be fused together (e.g. with
+      // Eigen).
+      at::native::add_out(prealloc_tensors[0], wide, mu_);
+      at::native::mul_out(prealloc_tensors[1], prealloc_tensors[0], sigma_);
+
+      at::native::clamp_out(
+          prealloc_tensors[2], prealloc_tensors[1], -10.0, 10.0);
+
+      // Potential optimization: original tensor could be pre-transposed.
+      // prealloc_tensors[3] = at::native::transpose(user_emb, 1, 2);
+      auto sizes = user_emb.sizes();
+      auto strides = user_emb.strides();
+      prealloc_tensors[3].set_(
```

Comment:
yeah, it seems correct - maybe because we still call `set_` and go through a bunch of field assignments? (even though TensorImpl is the same). You can probably wrap it into the same `if (prealloc_tensors[3].data_ptr() != user_emb.data_ptr())` as below and see whether it helps